File: CloudSeed/ModulatedDelay.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using AudioLib;

namespace CloudSeed
{
	public class ModulatedDelay
	{
		private const int ModulationUpdateRate = 8;

		private readonly double[] buffer;
		private readonly double[] output;
		private readonly int bufferSize;
		private int writeIndex;
		private int readIndexA;
		private int readIndexB;
		private int samplesProcessed;

		private double modPhase;
		private double gainA;
		private double gainB;

		public int SampleDelay;

		public double ModAmount;
		public double ModRate;
		
		public ModulatedDelay(int bufferSize, int sampleDelay)
		{
			this.bufferSize = bufferSize;
			this.buffer = new double[bufferSize];
			this.output = new double[bufferSize];
			this.SampleDelay = sampleDelay;
			writeIndex = 0;
			Update();
		}

		public double[] Output { get { return output; } }

		public void Process(double[] input, int sampleCount)
		{
			for (int i = 0; i < sampleCount; i++)
			{
				if (samplesProcessed == ModulationUpdateRate)
					Update();

				buffer[writeIndex] = input[i];
				output[i] = buffer[readIndexA] * gainA + buffer[readIndexB] * gainB;

				writeIndex++;
				readIndexA++;
				readIndexB++;
				if (writeIndex >= bufferSize) writeIndex -= bufferSize;
				if (readIndexA >= bufferSize) readIndexA -= bufferSize;
				if (readIndexB >= bufferSize) readIndexB -= bufferSize;
				samplesProcessed++;
			}
		}

		public void ClearBuffers()
		{
			buffer.Zero();
		}

		private void Update()
		{
			modPhase += ModRate * ModulationUpdateRate;
			if (modPhase > 1) modPhase -= 1;

			var mod = FastSin.Get(modPhase);
			var totalDelay = SampleDelay + ModAmount * mod;

			var delayA = (int)totalDelay;
			var delayB = (int)totalDelay + 1;

			var partial = totalDelay - delayA;

			gainA = 1 - partial;
			gainB = partial;
			
			readIndexA = writeIndex - delayA;
			readIndexB = writeIndex - delayB;
			if (readIndexA < 0) readIndexA += bufferSize;
			if (readIndexB < 0) readIndexB += bufferSize;

			samplesProcessed = 0;
		}
	}
}

File: CloudSeed.Native/ModulatedDelay.cpp
#include "ModulatedDelay.h"
#include "Utils.h"
#include "FastSin.h"

namespace CloudSeed
{
	ModulatedDelay::ModulatedDelay(int bufferSize, int sampleDelay)
	{
		this->bufferSize = bufferSize;
		this->delayBuffer = new double[DelayBufferSamples];
		this->output = new double[bufferSize];
		this->SampleDelay = sampleDelay;
		writeIndex = 0;
		modPhase = 0.01 + 0.98 * (std::rand() / (double)RAND_MAX);
		ModRate = 0.0;
		ModAmount = 0.0;
		Update();
	}

	ModulatedDelay::~ModulatedDelay()
	{
		delete delayBuffer;
		delete output;
	}

	double* ModulatedDelay::GetOutput()
	{
		return output;
	}

	void ModulatedDelay::Process(double* input, int sampleCount)
	{
		for (int i = 0; i < sampleCount; i++)
		{
			if (samplesProcessed == ModulationUpdateRate)
				Update();

			delayBuffer[writeIndex] = input[i];
			output[i] = delayBuffer[readIndexA] * gainA + delayBuffer[readIndexB] * gainB;

			writeIndex++;
			readIndexA++;
			readIndexB++;
			if (writeIndex >= DelayBufferSamples) writeIndex -= DelayBufferSamples;
			if (readIndexA >= DelayBufferSamples) readIndexA -= DelayBufferSamples;
			if (readIndexB >= DelayBufferSamples) readIndexB -= DelayBufferSamples;
			samplesProcessed++;
		}
	}

	void ModulatedDelay::ClearBuffers()
	{
		Utils::ZeroBuffer(delayBuffer, DelayBufferSamples);
		Utils::ZeroBuffer(output, bufferSize);
	}

	void ModulatedDelay::Update()
	{
		modPhase += ModRate * ModulationUpdateRate;
		if (modPhase > 1) 
			modPhase = std::fmod(modPhase, 1.0);

		auto mod = FastSin::Get(modPhase);
		auto totalDelay = SampleDelay + ModAmount * mod;

		auto delayA = (int)totalDelay;
		auto delayB = (int)totalDelay + 1;

		auto partial = totalDelay - delayA;

		gainA = 1 - partial;
		gainB = partial;

		readIndexA = writeIndex - delayA;
		readIndexB = writeIndex - delayB;
		if (readIndexA < 0) readIndexA += DelayBufferSamples;
		if (readIndexB < 0) readIndexB += DelayBufferSamples;

		samplesProcessed = 0;
	}
}
